openhands/runtime/utils/runtime_build.py

```diff
@@ -224,160 +117,181 @@ def build_runtime_image(
+    project_root = openhands_source_dir.parent
+    logger.info(f'Building source distribution using project root: {project_root}')
+
+    # Copy the 'openhands' directory (Source code)
+    shutil.copytree(
+        openhands_source_dir,
+        Path(build_folder, 'code', 'openhands'),
+        ignore=shutil.ignore_patterns(
+            '.*/',
+            '__pycache__/',
+            '*.pyc',
+        ),
+    )
 
-    Parameters:
-    - docker_folder (str): the path to the docker build folder
-    - runtime_builder (RuntimeBuilder): the runtime builder instance
-    - target_image_repo (str): the repository name for the target image
-    - target_image_hash_tag (str): the *hash* tag for the target image that is calculated based
-        on the contents of the docker build folder (source code and Dockerfile)
-        e.g. 1234567890abcdef
-    -target_image_tag (str): the tag for the target image that's generic and based on the base image name
-        e.g. oh_v0.9.3_image_ubuntu_tag_22.04
-    """
-    target_image_hash_name = f'{target_image_repo}:{target_image_hash_tag}'
-    target_image_generic_name = f'{target_image_repo}:{target_image_tag}'
+    # Copy pyproject.toml and poetry.lock files
+    for file in ['pyproject.toml', 'poetry.lock']:
+        src = Path(openhands_source_dir, file)
+        if not src.exists():
+            src = Path(project_root, file)
+        shutil.copy2(src, Path(build_folder, 'code', file))
 
-    tags_to_add = [target_image_hash_name]
+    # Create a Dockerfile and write it to build_folder
+    dockerfile_content = _generate_dockerfile(
+        base_image,
+        build_from_scratch=build_from_scratch,
+        extra_deps=extra_deps,
+    )
+    with open(Path(build_folder, 'Dockerfile'), 'w') as file:  # type: ignore
+        file.write(dockerfile_content)  # type: ignore
+
+
+def get_hash_for_lock_files(base_image: str):
+    openhands_source_dir = Path(openhands.__file__).parent
+    md5 = hashlib.md5()
+    md5.update(base_image.encode())
+    for file in ['pyproject.toml', 'poetry.lock']:
+        src = Path(openhands_source_dir, file)
+        if not src.exists():
+            src = Path(openhands_source_dir.parent, file)
+        with open(src, 'rb') as f:
+            for chunk in iter(lambda: f.read(4096), b''):
+                md5.update(chunk)
+    # We get away with truncation because we want something that is unique
+    # rather than something that is cryptographically secure
+    result = md5.hexdigest()[:16]
+    return result
+
+
+def get_hash_for_source_files():
+    openhands_source_dir = Path(openhands.__file__).parent
+    dir_hash = dirhash(
+        openhands_source_dir,
+        'md5',
+        ignore=[
+            '.*/',  # hidden directories
+            '__pycache__/',
+            '*.pyc',
+        ],
+    )
+    # We get away with truncation because we want something that is unique
+    # rather than something that is cryptographically secure
+    result = dir_hash[:16]
```

Comment:
ditto - maybe longer with 32?

Reply:
Ok - I updated the hash to use base36 truncated to 16 characters. This gives it a collision space of 1:22,300,745,198,530,623,141,535,718,272,648,361,505,980,416